File: packages/sqlite/native/Sqlite.swift
import Foundation

/// Process-wide open database count for CI in-memory stubs.
enum LucentSqliteCounts {
  private static let lock = NSLock()
  private static var databases = 0

  static func adjustDatabases(_ delta: Int) {
    lock.lock()
    databases += delta
    lock.unlock()
  }

  static var liveDatabases: Double {
    lock.lock()
    defer { lock.unlock() }
    return Double(databases)
  }
}

/// Borrowed transaction handle — valid only for the duration of `transaction`
/// (`retention: "call"`). Invalidated when the body returns.
final class LucentSQLiteTransaction: @unchecked Sendable {
  private let database: LucentSQLiteDatabase
  private var valid = true

  init(database: LucentSQLiteDatabase) {
    self.database = database
  }

  func invalidate() {
    valid = false
  }

  func execute(sql: String, params: [String]) throws {
    guard valid else { throw LucentError(code: "TX_CLOSED") }
    try database.execute(sql: sql, params: params)
  }

  func query(sql: String, params: [String]) throws -> Double {
    guard valid else { throw LucentError(code: "TX_CLOSED") }
    return try database.query(sql: sql, params: params)
  }
}

/// CI SQLite stub backed by an in-memory table map with real rollback on throw.
final class LucentSQLiteDatabase: @unchecked Sendable {
  private let lock = NSLock()
  private var isClosed = false
  private var store: [String: [String]] = [:]
  private var snapshot: [String: [String]]?

  private init() {
    LucentSqliteCounts.adjustDatabases(1)
  }

  static func open(path: String) throws -> LucentSQLiteDatabase {
    _ = path
    return LucentSQLiteDatabase()
  }

  var closed: Bool {
    lock.lock()
    defer { lock.unlock() }
    return isClosed
  }

  func execute(sql: String, params: [String]) throws {
    lock.lock()
    defer { lock.unlock() }
    guard !isClosed else { throw LucentError(code: "CLOSED") }
    if sql.uppercased().hasPrefix("THROW") {
      throw LucentError(code: "TX_FORCE_FAIL")
    }
    let key = params.first ?? sql
    if sql.uppercased().hasPrefix("DELETE") {
      store.removeValue(forKey: key)
    } else {
      store[key] = params
    }
  }

  /// Returns matching row count for the CI map.
  func query(sql: String, params: [String]) throws -> Double {
    lock.lock()
    defer { lock.unlock() }
    guard !isClosed else { throw LucentError(code: "CLOSED") }
    let key = params.first ?? sql
    return store[key] == nil ? 0 : 1
  }

  /// Body receives a borrowed tx; commit on success, restore snapshot on throw.
  func transaction(body: (LucentSQLiteTransaction) throws -> Void) throws {
    lock.lock()
    guard !isClosed else {
      lock.unlock()
      throw LucentError(code: "CLOSED")
    }
    snapshot = store
    lock.unlock()
    let tx = LucentSQLiteTransaction(database: self)
    defer { tx.invalidate() }
    do {
      try body(tx)
      lock.lock()
      snapshot = nil
      lock.unlock()
    } catch {
      lock.lock()
      if let snap = snapshot { store = snap }
      snapshot = nil
      lock.unlock()
      throw error
    }
  }

  func close() async {
    await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
      lock.lock()
      let release = !isClosed
      if release { isClosed = true }
      lock.unlock()
      if release { LucentSqliteCounts.adjustDatabases(-1) }
      continuation.resume()
    }
  }
}

File: packages/sqlite/native/Sqlite.kt
package {{androidPackage}}

import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

/** Process-wide open database count for CI in-memory stubs. */
object LucentSqliteCounts {
  private val lock = Any()
  private var databases = 0

  fun adjustDatabases(delta: Int) = synchronized(lock) { databases += delta }

  val liveDatabases: Double get() = synchronized(lock) { databases.toDouble() }
}

/**
 * Borrowed transaction handle — valid only for the duration of `transaction`
 * (`retention: "call"`). Invalidated when the body returns.
 */
class LucentSQLiteTransaction(private val database: LucentSQLiteDatabase) {
  private var valid = true

  fun invalidate() {
    valid = false
  }

  fun execute(sql: String, params: MutableList<String>) {
    if (!valid) throw LucentError("TX_CLOSED", "Transaction is no longer valid")
    database.execute(sql, params)
  }

  fun query(sql: String, params: MutableList<String>): Double {
    if (!valid) throw LucentError("TX_CLOSED", "Transaction is no longer valid")
    return database.query(sql, params)
  }
}

/** CI SQLite stub backed by an in-memory table map with real rollback on throw. */
class LucentSQLiteDatabase private constructor() {
  private val lock = Any()
  private var isClosed = false
  private val store = mutableMapOf<String, MutableList<String>>()
  private var snapshot: Map<String, MutableList<String>>? = null

  init {
    LucentSqliteCounts.adjustDatabases(1)
  }

  companion object {
    fun open(path: String): LucentSQLiteDatabase {
      path.length // path reserved for real SQLite
      return LucentSQLiteDatabase()
    }
  }

  val closed: Boolean
    get() = synchronized(lock) { isClosed }

  fun execute(sql: String, params: MutableList<String>) {
    synchronized(lock) {
      if (isClosed) throw LucentError("CLOSED", "Database is closed")
      if (sql.uppercase().startsWith("THROW")) throw LucentError("TX_FORCE_FAIL", "Forced transaction failure")
      val key = params.firstOrNull() ?: sql
      if (sql.uppercase().startsWith("DELETE")) store.remove(key)
      else store[key] = params.toMutableList()
    }
  }

  /** Returns matching row count for the CI map. */
  fun query(sql: String, params: MutableList<String>): Double {
    synchronized(lock) {
      if (isClosed) throw LucentError("CLOSED", "Database is closed")
      val key = params.firstOrNull() ?: sql
      return if (store.containsKey(key)) 1.0 else 0.0
    }
  }

  /** Body receives a borrowed tx; commit on success, restore snapshot on throw. */
  fun transaction(body: (LucentSQLiteTransaction) -> Unit) {
    synchronized(lock) {
      if (isClosed) throw LucentError("CLOSED", "Database is closed")
      snapshot = store.mapValues { it.value.toMutableList() }
    }
    val tx = LucentSQLiteTransaction(this)
    try {
      body(tx)
      synchronized(lock) { snapshot = null }
    } catch (error: Throwable) {
      synchronized(lock) {
        snapshot?.let { snap ->
          store.clear()
          store.putAll(snap.mapValues { it.value.toMutableList() })
        }
        snapshot = null
      }
      throw error
    } finally {
      tx.invalidate()
    }
  }

  suspend fun close(): Unit = suspendCoroutine { continuation ->
    val release = synchronized(lock) {
      if (isClosed) false
      else {
        isClosed = true
        true
      }
    }
    if (release) LucentSqliteCounts.adjustDatabases(-1)
    continuation.resume(Unit)
  }
}
